experimental/algorithm/LAGr_SwapEdges.c

```
@@ -0,0 +1,822 @@
+        GRB_TRY (GrB_free(&(con->b))) ;
+        con->b = dup_swaps_v;
+        con->nvals = n_keep;
+        con->format = GxB_BITMAP;
+        dup_swaps_v = NULL;
+        GRB_TRY (GxB_load_Vector_from_Container(M, con, NULL)) ;
+        GRB_TRY (GrB_free(&con)) ;
+        GRB_TRY (LAGraph_FastAssign_Semiring(
+            E_vec, NULL, second_edge, edge_perm, M, ramp_v, 
+            second_second_edge, NULL, msg)) ;
+        
+
+        n_keep /= 2;
+
+        FREE_LOOP ; // Free Matricies that have to be rebuilt
+
+        num_swaps += n_keep ;
+        LG_TRY (LAGraph_Random_Next(random_v, msg)) ;
+        printf("Made %ld swaps this loop. "
+                "[%.3f%% of Planned, %.3f%% of edges swapped]\n"
+                "Completed %ld swaps total. [%.3f%% of Planned]\n", 
+             n_keep, n_keep * 100.0 / totSwaps, n_keep * 200.0 / e, num_swaps, 
+             num_swaps * 100.0 / totSwaps) ;
+        if(n_keep < (int) (loopMin * e / 2) + 1)
+        {
+            printf("Too Few Swaps occured! Exiting.\n");
+            break;
+        }
+    }
+    GRB_TRY (GxB_Vector_unload(
+        E_vec, (void **) &indices, &lg_edge, &e, &ind_size, &E_hand, NULL));
+    GRB_TRY (GxB_Vector_load(
+        E_vec, (void **) &indices, E_type, e * 2, ind_size, E_hand, NULL));
+    GRB_TRY (GrB_Vector_extract(
+        Aj, NULL, NULL, E_vec, stride, GxB_STRIDE, NULL));
+    stride[GxB_BEGIN] = 0;
+    GRB_TRY (GrB_Vector_extract(
+        Ai, NULL, NULL, E_vec, stride, GxB_STRIDE, NULL));
+    // Build Output Matrix
+    GRB_TRY (GrB_Matrix_new(&A_new, GrB_BOOL, n, n)) ;
+    GRB_TRY (GxB_Matrix_build_Scalar_Vector(A_new, Ai, Aj, one8, NULL)) ;
+    GRB_TRY (GrB_eWiseAdd(
+        A_new, NULL, NULL, GrB_LOR_MONOID_BOOL, A_new,A_new, GrB_DESC_T0
+    )) ;
+    LAGRAPH_TRY (LAGraph_New (G_new, &A_new, LAGraph_ADJACENCY_DIRECTED, msg)) ;
+    LG_FREE_WORK ;
+    return (num_swaps >= totSwaps)? GrB_SUCCESS :  LAGRAPH_INSUFFICIENT_SWAPS ;
```

I defined LAGRAPH_INSUFFICIENT_SWAPS as 2100 in LAGraphX.h. Was this correct?
Should I return the number of swaps I made? As its own output variable, or do I make totSwaps input/output?

---

I'm reluctant to add a new enum value to the LAGraph return values that is unique to just one algorithm.  How about returning LAGRAPH_CONVERGENCE_FAILURE instead?  The method iterates, trying to find swaps, and if it doesn't succeed, it does need to return an error.

The description of LAGRAPH_CONVERGENCE_FAILURE is quite similar,  "An iterative process failed to converge to a good solution".

What do you think?

---

I think returning an error is reasonable. Although currently, I return the error, but still return a valid G_new graph with however many swaps I managed to get before going under the per-loop threshold. Is this a reasonable thing to do, or does returning an error imply the output graph should be freed?

---

I see.  Maybe we should talk it over.  In your current design, LAGRAPH_INSUFFICIENT_SWAPS is a warning, not an error, right?